src/BloomExe/Book/BookCompressor.cs

```diff
@@ -198,42 +203,59 @@ private static void MakeExtraEntry(ZipOutputStream zipStream, string name, strin
-			// those spots.
-			// Note that the src path can end either with the closing quote of the src attr or with a
-			// question mark, signifying options (like ?optional=true for branding images) which are
-			// not part of the file we want to search for.
-			var regex = new Regex("<img[^>]*src\\s*=\\s*(['\"])(.*?)(\\1|\\?)[^>]*(/>|>\\s*</img\\s*>)");
-			var match = regex.Match(content);
 			var folderPath = Path.GetDirectoryName(bookFile);
-			while (match.Success)
+			foreach (var imgElt in dom.SafeSelectNodes("//img[@src]").Cast<XmlElement>().ToArray())
 			{
-				var file = match.Groups[2].Value;
+				var file = imgElt.Attributes["src"].Value.Split('?')[0];
 				if (!File.Exists(Path.Combine(folderPath, file)))
 				{
-					content = content.Substring(0, match.Index) + content.Substring(match.Index + match.Length);
-					match = regex.Match(content, match.Index);
-				}
-				else
-				{
-					match = regex.Match(content, match.Index + match.Length);
+					imgElt.ParentNode.RemoveChild(imgElt);
 				}
 			}
-			return content;
 		}
 
-		private static string StripContentEditable(string input)
+		private static void StripContentEditable(XmlDocument dom)
 		{
-			var regex = new Regex("\\s*contenteditable\\s*=\\s*(['\"]).*?\\1");
-			return regex.Replace(input, "");
+			foreach (var editableElt in dom.SafeSelectNodes("//div[@contenteditable]").Cast<XmlElement>().ToArray())
+			{
+				editableElt.RemoveAttribute("contenteditable");
+			}
+		}
+
+		/// <summary>
+		/// Find every place in the html file where an img element is nested inside a div with class bloom-imageContainer.
+		/// Convert the img into a background image of the image container div.
+		/// Specifically, make the following changes:
+		/// - Copy any data-x attributes from the img element to the div
+		/// - Convert the src attribute of the img to style="background-image:url('...')" (with the same source) on the div
+		///    (any pre-existing style attribute on the div is lost)
+		/// - Add the class bloom-backgroundImage to the div
+		/// - delete the img element
+		/// (See oldImg and newImg in unit test CompressBookForDevice_ImgInImgContainer_ConvertedToBackground for an example).
+		/// </summary>
+		/// <param name="wholeBookHtml"></param>
+		/// <returns></returns>
+		private static void ConvertImagesToBackground(XmlDocument dom)
+		{
+			foreach (var imgContainer in dom.SafeSelectNodes("//div[contains(@class, 'bloom-imageContainer')]").Cast<XmlElement>().ToArray())
+			{
+				var img = imgContainer.ChildNodes.Cast<XmlNode>().FirstOrDefault(n => n is XmlElement && n.Name == "img");
+				if (img == null || img.Attributes["src"] == null)
+					continue;
+				// The filename shouldn't have double quotes because that's not valid (at least in Windows).
+				// It might have single quotes, which need escaping because we use them to delimit the url value in the background-image.
+				var src = img.Attributes["src"].Value.Replace("'", @"\'");
```

Comment:
I wonder if it would be appropriate to use HtmlDom.SetImageElementUrl() and to make sure that it contains any of this quote logic that appears to be unique here? Though I'm not clear how this late in the game the images could have problematic names.

Also I've tried to make everything that deals with urls use UrlPathString, so that we don't have logic spread around and we reduce those pesky errors related to double encoding/decoding, etc. Would be nice to see that continued here.